client/prompt.lua

```diff
@@ -26,25 +25,20 @@ local function startPrompt(promptTable)
 
   promptIsOpen = true
 
-  RegisterNuiCallbackType(cbStr)
-  local eventData = AddEventHandler('__cfx_nui:' .. cbStr, function(data, cb)
-    p:resolve(data)
+  RegisterRawNuiCallback(cbStr, function(data, cb)
+    local resp = json.decode(data.body)
+
+    p:resolve(resp)
     promptIsOpen = false
-    cb({})
+    cb({ body = json.encode({}) })
+
+    UnregisterRawNuiCallback(`promptResp-${props.id}`);
   end)
 
   local result = Citizen.Await(p)
-  -- TODO: This causes a memory leak at the current moment
-  -- as the RegisteredType is never cleaned up from the ResourceUI
-  -- container in cpp. Feature pending
-
-  -- Before we remove the event handler we need to copy the data
-  -- otherwise we won't get a proper response.
-  local resultCopy = result
 
-  RemoveEventHandler(eventData)
   -- Returns two results as ['closed' | 'submitted', content | null]
-  return resultCopy[1], resultCopy[2]
+  return result[1], result[2]
```

Comment:
Not sure if it’s really needed but if you’re going to await a result it may be a good idea to do some sort of timeout for when the promise isn’t resolved within a certain time frame.